packages/lit-html/src/directives/async-replace.ts

```diff
@@ -7,72 +7,82 @@
 import {ChildPart, noChange} from '../lit-html.js';
 import {directive, DirectiveParameters} from '../directive.js';
 import {AsyncDirective} from '../async-directive.js';
+import {DisconnectableAwaiter} from './disconnectable-awaiter.js';
 
 type Mapper<T> = (v: T, index?: number) => unknown;
 
-class AsyncReplaceDirective extends AsyncDirective {
-  private _value?: AsyncIterable<unknown>;
-  private _reconnectResolver?: () => void;
-  private _reconnectPromise?: Promise<void>;
+export class AsyncReplaceDirective extends AsyncDirective {
+  private __iterable!: AsyncIterable<unknown>;
+  private __iterator!: AsyncIterator<unknown>;
+  protected index = 0;
+  private __nextAwaiter?: DisconnectableAwaiter<this, IteratorResult<unknown>>;
+  private __mapper?: Mapper<unknown>;
 
   // @ts-expect-error value not used, but we want a nice parameter for docs
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   render<T>(value: AsyncIterable<T>, _mapper?: Mapper<T>) {
     return noChange;
   }
 
-  update(_part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+  update(_part: ChildPart, [iterable, mapper]: DirectiveParameters<this>) {
     // If we've already set up this particular iterable, we don't need
     // to do anything.
-    if (value === this._value) {
+    if (iterable === this.__iterable) {
       return;
     }
-    this._value = value;
-    this.__iterate(mapper);
+    // If we're starting a new promise, stop tracking the last one, since its
+    // result is no longer relevant
+    this.__nextAwaiter?.disconnect();
+    this.__mapper = mapper;
+    this.__iterable = iterable;
+    this.index = 0;
+    // Note, we use the iterator protocol manually (rather than for/await/of),
+    // which lets us use the `DisconnectableAwaiter` helper to disconnect &
+    // reconnect awaiting the next value from the iterable
+    this.__iterator = iterable[Symbol.asyncIterator]();
```

Comment:
Nit: This is a slight change it behavior. `for await` will consume a sync iterator if the async iterator isn't define.

Reply:
Reverted to `for await`